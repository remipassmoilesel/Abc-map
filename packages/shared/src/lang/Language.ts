/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

/**
 * Supported languages
 *
 * Values must be compatible with lang attribute: https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/lang
 */
export enum Language {
  English = 'en',
  French = 'fr',
}

export const SupportedLanguages = Object.values(Language);

export const FallbackLang = Language.English;

export function isLangSupported(lang: string | Language): boolean {
  return Object.values(Language).includes(lang as Language);
}

export function langFromString(lang: string): Language | undefined {
  return Object.values(Language).find((l) => l === lang);
}
