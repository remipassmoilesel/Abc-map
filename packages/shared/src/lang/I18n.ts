/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { FallbackLang, Language } from './Language';

export interface I18nText {
  language: Language;
  text: string;
}

export interface I18nList {
  language: Language;
  text: string[];
}

export function getTextByLang(texts: I18nText[], lang: Language | string): string | undefined {
  const wanted = texts.find((desc) => desc.language === lang);
  const fallback = texts.find((desc) => desc.language === FallbackLang);
  const first = texts.length ? texts[0] : undefined;
  return wanted?.text ?? fallback?.text ?? first?.text;
}

export function getListByLang(texts: I18nList[], lang: Language | string): string[] | undefined {
  const wanted = texts.find((desc) => desc.language === lang);
  const fallback = texts.find((desc) => desc.language === FallbackLang);
  const first = texts.length ? texts[0] : undefined;
  return wanted?.text ?? fallback?.text ?? first?.text;
}
