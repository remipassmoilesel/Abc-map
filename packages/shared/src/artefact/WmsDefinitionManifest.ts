/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { AbcProjection } from '../project';
import { LayerAuthentication } from '../project/layer/LayerAuthentication';
import { I18nText } from '../lang';

export interface WmsDefinitionManifest {
  version: string;
  wms: {
    urls: string[];
    remoteLayerName: string;
    projection?: AbcProjection;
    extent?: [number, number, number, number];
    auth?: LayerAuthentication;
    prompt?: {
      name: string;
      type: 'string' | 'number';
      description: I18nText[];
    }[];
  };
}
