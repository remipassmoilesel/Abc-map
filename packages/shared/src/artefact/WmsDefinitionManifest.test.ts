/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */
import { Language } from '../lang';
import { WmsDefinitionManifest } from './WmsDefinitionManifest';

describe('WmsDefinitionManifest', () => {
  it('should not change without datastore migration', () => {
    const manifest: WmsDefinitionManifest = {
      version: 'version',
      wms: {
        urls: ['urls'],
        remoteLayerName: 'remoteLayerName',
        projection: { name: 'projection' },
        extent: [1, 2, 3, 4],
        auth: {
          username: 'username',
          password: 'password',
        },
        prompt: [
          {
            name: 'name',
            type: 'string',
            description: [
              { language: Language.French, text: 'description' },
              { language: Language.English, text: 'description' },
            ],
          },
        ],
      },
    };

    expect(manifest).toMatchSnapshot();
  });
});
