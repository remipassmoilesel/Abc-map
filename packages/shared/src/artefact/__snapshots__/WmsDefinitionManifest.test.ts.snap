// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WmsDefinitionManifest should not change without datastore migration 1`] = `
Object {
  "version": "version",
  "wms": Object {
    "auth": Object {
      "password": "password",
      "username": "username",
    },
    "extent": Array [
      1,
      2,
      3,
      4,
    ],
    "projection": Object {
      "name": "projection",
    },
    "prompt": Array [
      Object {
        "description": Array [
          Object {
            "language": "fr",
            "text": "description",
          },
          Object {
            "language": "en",
            "text": "description",
          },
        ],
        "name": "name",
        "type": "string",
      },
    ],
    "remoteLayerName": "remoteLayerName",
    "urls": Array [
      "urls",
    ],
  },
}
`;
